mm2src/mm2_event_stream/src/manager.rs

```diff
@@ -0,0 +1,516 @@
+use futures::channel::oneshot;
+use parking_lot::{RwLock, RwLockReadGuard, RwLockWriteGuard};
+use tokio::sync::mpsc;
+
+/// The errors that could originate from the streaming manager.
+#[derive(Debug)]
+pub enum StreamingManagerError {
+    /// There is no streamer with the given ID.
+    StreamerNotFound,
+    /// Couldn't send the data to the streamer.
+    SendError(String),
+    /// The streamer doesn't accept an input.
+    NoDataIn,
+    /// Couldn't spawn the streamer.
+    SpawnError(String),
+    /// The client is not known/registered.
+    UnknownClient,
+    /// A client with the same ID already exists.
+    ClientExists,
+    /// The client is already listening to the streamer.
+    ClientAlreadyListening,
+}
+
+struct StreamerInfo {
+    /// The communication channel to the streamer.
+    data_in: Option<UnboundedSender<Box<dyn Any + Send>>>,
+    /// Clients the streamer is serving for.
+    clients: HashSet<u64>,
+    /// The shutdown handle of the streamer.
+    shutdown: oneshot::Sender<()>,
+}
+
+impl StreamerInfo {
+    fn new(data_in: Option<UnboundedSender<Box<dyn Any + Send>>>, shutdown: oneshot::Sender<()>) -> Self {
+        Self {
+            data_in,
+            clients: HashSet::new(),
+            shutdown,
+        }
+    }
+
+    fn add_client(&mut self, client_id: u64) { self.clients.insert(client_id); }
+
+    fn remove_client(&mut self, client_id: &u64) { self.clients.remove(client_id); }
+
+    fn is_down(&self) -> bool { self.shutdown.is_canceled() }
+}
+
+struct ClientInfo {
+    /// The streamers the client is listening to.
+    listening_to: HashSet<String>,
+    /// The communication/stream-out channel to the client.
+    // NOTE: Here we are using `tokio`'s `mpsc` because the one in `futures` have some extra feature
+    // (ref: https://users.rust-lang.org/t/why-does-try-send-from-crate-futures-require-mut-self/100389).
+    // This feature is aimed towards the multi-producer case (which we don't use) and requires a mutable
+    // reference on `try_send` calls. This will require us to put the channel in a mutex and degrade the
+    // broadcasting performance.
+    channel: mpsc::Sender<Arc<Event>>,
+}
+
+impl ClientInfo {
+    fn new(channel: mpsc::Sender<Arc<Event>>) -> Self {
+        Self {
+            listening_to: HashSet::new(),
+            channel,
+        }
+    }
+
+    fn add_streamer(&mut self, streamer_id: String) { self.listening_to.insert(streamer_id); }
+
+    fn remove_streamer(&mut self, streamer_id: &str) { self.listening_to.remove(streamer_id); }
+
+    fn listens_to(&self, streamer_id: &str) -> bool { self.listening_to.contains(streamer_id) }
+
+    fn send_event(&self, event: Arc<Event>) {
+        // Only `try_send` here. If the channel is full (client is slow), the message
+        // will be dropped and the client won't receive it.
+        // This avoids blocking the broadcast to other receivers.
+        self.channel.try_send(event).ok();
+    }
```

Comment:
so it was 1024 before but with `.await` on send. i think `1024` is more than enough but we shouldn't `.await` for performance reasons.
imo having a backlog of 1024 unpolled events by the client is already very slow that we could care about messages disappearing, not to mention every `arc<event` not being consumed out of this channel means that we have some lone reference to a potentially large event object with large message inside, so not favoring the unbounded route here in a place where the other end of the connection has that much control.

Reply:
I don't have an opinion on the optimal channel size at the moment (I don't know the message frequency with all these new changes), but I would at least print the error message if sending fails, so we can debug it in the logs. Right now, when a disappearing message causes some logic to behave unusually it will be incredibly hard to find the reason.